cpp/src/arrow/c/abi.h

```diff
@@ -106,6 +169,77 @@ struct ArrowArrayStream {
 
 #endif  // ARROW_C_STREAM_INTERFACE
 
+#ifndef ARROW_C_DEVICE_STREAM_INTERFACE
+#define ARROW_C_DEVICE_STREAM_INTERFACE
+
+struct ArrowDeviceArrayStream {
+  // The device that this stream produces data on.
+  // All ArrowDeviceArrays that are produced by this
+  // stream should have the same device_type as set
+  // here. The device_type needs to be provided here
+  // so that consumers can provide the correct type
+  // of stream_ptr when calling get_next.
+  ArrowDeviceType device_type;
+
+  // Callback to get the stream schema
+  // (will be the same for all arrays in the stream).
+  //
+  // Return value: 0 if successful, an `errno`-compatible error code otherwise.
+  //
+  // If successful, the ArrowSchema must be released independently from the stream.
+  int (*get_schema)(struct ArrowDeviceArrayStream*, struct ArrowSchema* out);
+
+  // Callback to get the device id for the next array.
+  // This is necessary so that the proper/correct stream pointer can be provided
+  // to get_next. The parameter provided must not be null.
+  //
+  // Return value: 0 if successful, an `errno`-compatible error code otherwise.
+  //
+  // The next call to `get_next` should provide an ArrowDeviceArray whose
+  // device_id matches what is provided here, and whose device_type is the
+  // same as the device_type member of this stream.
+  int (*get_next_device_id)(struct ArrowDeviceArrayStream*, int* out_device_id);
+
+  // Callback to get the next array
+  // (if no error and the array is released, the stream has ended)
+  //
+  // the provided stream_ptr should be the appropriate stream, or
+  // equivalent object, for the device that the data is allocated on
+  // to indicate where the consumer wants the data to be accessible.
+  // if stream_ptr is NULL then the default stream (e.g. CUDA stream 0)
+  // should be used to ensure that the memory is accessible from any stream.
```

Comment:
Does this make sense for all backends, or are we trying to encode device-specific behaviour in a generic API?

Reply:
As far as I'm aware, this should make sense for all backends. If they utilize the concept of a stream, they all have some "default" stream that can be used if this is null (the example given being stream 0 for CUDA), so I'm fairly certain this isn't device-specific behavior being encoded. @kkraus14 can you fact-check me on this?

Reply:
Also sidenote, it would be nice if we could find a less ambiguous term, as "stream" already refers to something else in this interface (though I have no concrete suggestion).

Reply:
CUDA (NVIDIA GPUs) and ROCm (AMD GPUs) both use the term "stream". The SYCL equivalent is an in-order "queue". I'm not familiar enough with other device programming models to know what their terminology is or if a single entity is enough here.

Reply:
what do we all think about using `queue` here then instead of `stream`? Is that enough to avoid the ambiguity?